controllers/apps/transformer_cluster_backup_policy.go

```diff
@@ -82,139 +67,162 @@ func (r *clusterBackupPolicyTransformer) Transform(ctx graph.TransformContext, d
-				// cluster custom object, such as enable cluster backup, set backup
-				// schedule, etc.
-				// We always prioritize the cluster backup configuration in the
-				// cluster object, so we need to merge the cluster backup configuration
-				// into the default backup schedule created by backup policy template
-				// if it exists.
-				if needMergeClusterBackup {
-					newBackupSchedule = r.mergeClusterBackup(comp, backupPolicy, newBackupSchedule)
-				}
-				if newBackupSchedule == nil {
-					return
-				}
-				// if exist multiple backup policy templates and duplicate spec.identifier,
-				// the backupSchedule that may be generated may have duplicate names,
-				// and it is necessary to check if it already exists.
-				if _, ok := backupScheduleNames[newBackupSchedule.Name]; ok {
-					return
-				}
+	transformBackupPolicyAndSchedule := func(bpt *dpv1alpha1.BackupPolicyTemplate, compSpec *appsv1.ClusterComponentSpec, componentName string, isSharding, isHScaleTPL bool) error {
+		bpBuilder := newBackupPolicyBuilder(r, compSpec, bpt, componentName, isSharding, isHScaleTPL)
+		policy := transformBackupPolicy(bpBuilder)
+		// only merge the first backupSchedule for the cluster backup.
+		transformBackupSchedule(bpBuilder, policy)
+		return nil
+	}
 
-				if oldBackupSchedule == nil {
-					graphCli.Create(dag, newBackupSchedule)
-				} else {
-					graphCli.Patch(dag, oldBackupSchedule, newBackupSchedule)
-				}
-				graphCli.DependOn(dag, backupPolicy, newBackupSchedule)
-				comps := graphCli.FindAll(dag, &appsv1.Component{})
-				graphCli.DependOn(dag, backupPolicy, comps...)
-				backupScheduleNames[newBackupSchedule.Name] = struct{}{}
+	transformComponentBackupPolicy := func(compSpec *appsv1.ClusterComponentSpec, componentName string, isSharding bool) error {
+		compDef := r.ComponentDefs[compSpec.ComponentDef]
+		if compDef == nil {
+			return nil
+		}
+		hScaleBPTName := compDef.Annotations[constant.HorizontalScaleBackupPolicyTemplateKey]
+		if hScaleBPTName != "" {
+			bpt := &dpv1alpha1.BackupPolicyTemplate{}
+			if err := r.Client.Get(ctx.GetContext(), client.ObjectKey{Name: hScaleBPTName}, bpt); err != nil {
+				return err
 			}
-
-			// transform backup policy template to data protection backupPolicy
-			// and backupSchedule
-			compItems := r.getClusterComponentItems()
-			for j, v := range compItems {
-				policy := transformBackupPolicy(v)
-				// only merge the first backupSchedule for the cluster backup.
-				transformBackupSchedule(v, policy, j == 0)
+			if err := transformBackupPolicyAndSchedule(bpt, compSpec, componentName, isSharding, true); err != nil {
+				return err
 			}
 		}
+		bpt, err := r.getBackupPolicyTemplate(compSpec.ComponentDef, hScaleBPTName)
+		if err != nil {
+			return err
+		}
+		if bpt == nil {
+			return nil
+		}
+		return transformBackupPolicyAndSchedule(bpt, compSpec, componentName, isSharding, false)
+	}
+
+	for i := range r.Cluster.Spec.ComponentSpecs {
```

Comment:
Can we move all the logic into BackupPolicyTemplateReconciler?

Reply:
the backupPolicy and backupScheduled will reconcile with the cluster.spec.backup and componentSpecs/shardingSpecs.  if move it to BackupPolicyTemplateReconciler, bpt controller will watch the changes of the cluster and handle the logic for all related clusters .

Reply:
The backup policy needs a mechanism or the API to bind to the cluster in a loosely coupled way. For example, referring to Prometheus, the backup policy can define an annotation and well-defined value, so that the cluster only needs to serialize the necessary backup information to this annotation in the defined format, and the backup policy controller can generate the object for it automatically.

Reply:
When using a BPT (Backup Policy Template) controller, every time there is a change in the cluster, the corresponding BPT will reconcile all related clusters to synchronize the backup policies and schedules. If there are many clusters, this can keep the controller very busy, potentially causing blocks and delaying reconciliations. Prometheus, on the other hand, periodically pulls all metrics with the specified annotations, which is a reasonable approach.